// i18next-fs-backend/esm/writeFile.js
import JSON5 from './formats/json5.js';
import jsYaml from './formats/yaml.js';
import * as fsMod from './fs.cjs';
import extname from './extname.js';
var isDeno = typeof Deno !== 'undefined';
var YAML = typeof jsYaml !== 'undefined' && jsYaml.load ? jsYaml : undefined;
var fs = fsMod ? fsMod.default || fsMod : undefined;

var writeFileInNodeSync = function writeFileInNodeSync(filename, payload) {
  return fs.writeFileSync(filename, payload, 'utf8');
};

var writeFileInNode = function writeFileInNode(filename, payload) {
  return new Promise(function (resolve, reject) {
    return fs.writeFile(filename, payload, 'utf8', function (err, data) {
      return err ? reject(err) : resolve(data);
    });
  });
};

var writeFileInDenoSync = function writeFileInDenoSync(filename, payload) {
  var encoder = new TextEncoder();
  var data = encoder.encode(payload);
  Deno.writeFileSync(filename, data);
};

var writeFileInDeno = function writeFileInDeno(filename, payload) {
  var encoder = new TextEncoder();
  var data = encoder.encode(payload);
  return Deno.writeFile(filename, data);
};

var stringifyData = function stringifyData(extension, data, options) {
  var result = '';

  switch (extension) {
    case '.js':
    case '.ts':
      if (typeof module === 'undefined') {
        result = "export default ".concat(options.stringify(data, null, options.ident));
      } else {
        result = "module.exports = ".concat(options.stringify(data, null, options.ident));
      }

      break;

    case '.json5':
      result = JSON5.stringify(data, null, options.ident);
      break;

    case '.yml':
    case '.yaml':
      result = YAML.dump(data, {
        ident: options.indent
      });
      break;

    default:
      result = options.stringify(data, null, options.ident);
  }

  return result;
};

export function writeFileSync(filename, payload, options) {
  var ext = extname(filename);
  var data;

  try {
    data = stringifyData(ext, payload, options);
  } catch (err) {
    err.message = 'error stringifying ' + filename + ': ' + err.message;
    throw err;
  }

  if (isDeno) {
    return writeFileInDenoSync(filename, data);
  } else {
    return writeFileInNodeSync(filename, data);
  }
}
export function writeFile(filename, payload) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {
    stringify: JSON.stringify,
    ident: 2
  };
  var ext = extname(filename);
  var data;

  try {
    data = stringifyData(ext, payload, options);
  } catch (err) {
    err.message = 'error stringifying ' + filename + ': ' + err.message;
    throw err;
  }

  var fn = isDeno ? writeFileInDeno : writeFileInNode;
  return fn(filename, data);
}
